compiler/src/dmd/dsymbolsem.d

```diff
@@ -7412,3 +7412,132 @@ private extern(C++) class SetFieldOffsetVisitor : Visitor
         }
     }
 }
+
+Scope* newScope(Dsymbol d, Scope* sc)
+{
+    scope nsv = new newScopeVisitor(sc);
+    d.accept(nsv);
+    return nsv.sc;
+}
+
+extern(C++) class newScopeVisitor : Visitor
+{    
+    alias visit = typeof(super).visit;
+    Scope* sc;
+    this(Scope* sc)
+    {
+        this.sc = sc;
+    }
+
+override void visit(VisibilityDeclaration atbd)
+    {
+        if (atbd.pkg_identifiers)
+        {
+            dsymbolSemantic(atbd, sc);
+        }
+        sc = atbd.createNewScope(sc, sc.stc, sc.linkage, sc.cppmangle, atbd.visibility, 1, sc.aligndecl, sc.inlining);
+    }
+    /**
+     * Returns:
+     *   A copy of the parent scope, with `this` as `namespace` and C++ linkage
+     *///override Scope* visit(Scope* sc)
+    override void visit(CPPNamespaceDeclaration scd)
+    {   
+        auto scx = sc.copy();
+        scx.linkage = LINK.cpp;
+        scx.namespace = scd;
+        sc = scx;
+    }
+
+    override void visit(CPPMangleDeclaration cpmd)
+    {
+        sc = cpmd.createNewScope(sc, sc.stc, LINK.cpp, cpmd.cppmangle, sc.visibility, sc.explicitVisibility,
+            sc.aligndecl, sc.inlining);
+    }
+
+    override void visit(AlignDeclaration visd)
+    {
+     sc = visd.createNewScope(sc, sc.stc, sc.linkage, sc.cppmangle, sc.visibility, sc.explicitVisibility, visd, sc.inlining);
+    }
+
+    override void visit(PragmaDeclaration prd)
+    {
+        if (prd.ident == Id.Pinline)
+        {
+            // We keep track of this pragma inside scopes,
+            // then it's evaluated on demand in function semantic
+            sc = prd.createNewScope(sc, sc.stc, sc.linkage, sc.cppmangle, sc.visibility, sc.explicitVisibility, sc.aligndecl, prd); // @suppress(dscanner.style.long_line)
+        }
+    }
+
+override void visit(LinkDeclaration  lid)
+    {
+        sc= lid.createNewScope(sc, sc.stc, lid.linkage, sc.cppmangle, sc.visibility, sc.explicitVisibility,
+            sc.aligndecl, sc.inlining);
+    }
+
+/**
+     * Provides a new scope with `STC.deprecated_` and `Scope.depdecl` set
+     *
+     * Calls `StorageClassDeclaration.newScope` (as it must be called or copied
+     * in any function overriding `newScope`), then set the `Scope`'s depdecl.
+     *
+     * Returns:
+     *   Always a new scope, to use for this `DeprecatedDeclaration`'s members.
+     */
+    override void visit(DeprecatedDeclaration dpd)
+    {
+        auto oldsc = sc;
+        visit(cast(StorageClassDeclaration) dpd);
+        auto scx = sc;
+        sc = oldsc;
```

Comment:
Still need to add back in 
```
if (scx == sc)
     scx = sc.push();
scx.depdecl = did;
sc =  scx;
```

Reply:
this should be removed?
```
auto oldsc = sc;
        visit(cast(StorageClassDeclaration) dpd);
        auto scx = sc;
        sc = oldsc;
```

Reply:
no

Reply:
You mean something like this 
```
auto scx = dpd.newScope(sc);

         if (scx == sc)
        scx = sc.push();
        scx.depdecl = dpd;
        sc =  scx;
        auto oldsc = sc;
        visit(cast(StorageClassDeclaration) dpd);
        //auto scx = sc;
        sc = oldsc;
```

Reply:
```
auto oldsc = sc;
visit(cast(StorageClassDeclaration) dpd);
auto scx = sc;
sc = oldsc;
if (scx == sc) 
    scx = sc.push(); 
scx.depdecl = did; 
sc = scx;
```

Reply:
This code is confusing and error prone. I suggest we swallow the extra visitor call for the sake of sane code:

```d
        auto scx = (cast(StorageClassDeclaration)dpd).newScope(sc);
        // The enclosing scope is deprecated as well
        if (scx == sc)
            scx = sc.push();
        scx.depdecl = dpd;
        sc = scx;
```